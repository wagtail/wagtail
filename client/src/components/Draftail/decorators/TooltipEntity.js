import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Icon } from 'draftail';

import Tooltip from '../Tooltip/Tooltip';
import Portal from '../../Portal/Portal';

const shortenLabel = (label) => {
  let shortened = label;
  if (shortened.length > 25) {
    shortened = `${shortened.slice(0, 20)}â€¦`;
  }

  return shortened;
};

class TooltipEntity extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showTooltipAt: null,
    };

    this.onEdit = this.onEdit.bind(this);
    this.onRemove = this.onRemove.bind(this);
    this.openTooltip = this.openTooltip.bind(this);
    this.closeTooltip = this.closeTooltip.bind(this);
  }

  onEdit(e) {
    const { onEdit, entityKey } = this.props;

    e.preventDefault();
    e.stopPropagation();
    onEdit(entityKey);
  }

  onRemove(e) {
    const { onRemove, entityKey } = this.props;

    e.preventDefault();
    e.stopPropagation();
    onRemove(entityKey);
  }

  openTooltip(e) {
    const trigger = e.target.closest('[data-draftail-trigger]');

    // Click is within the tooltip.
    if (!trigger) {
      return;
    }

    const container = trigger.closest('[data-draftail-editor-wrapper]');
    const containerRect = container.getBoundingClientRect();
    const rect = trigger.getBoundingClientRect();

    this.setState({
      showTooltipAt: {
        container: container,
        top: rect.top - containerRect.top - (document.documentElement.scrollTop || document.body.scrollTop),
        left: rect.left - containerRect.left - (document.documentElement.scrollLeft || document.body.scrollLeft),
        width: rect.width,
        height: rect.height,
      },
    });
  }

  closeTooltip() {
    this.setState({ showTooltipAt: null });
  }

  render() {
    const {
      children,
      icon,
      label,
      url,
    } = this.props;
    const { showTooltipAt } = this.state;

    // Contrary to what JSX A11Y says, this should be a button but it shouldn't be focusable.
    /* eslint-disable springload/jsx-a11y/interactive-supports-focus */
    return (
      <a
        role="button"
        // Use onMouseUp to preserve focus in the text even after clicking.
        onMouseUp={this.openTooltip}
        className="TooltipEntity"
        data-draftail-trigger
      >
        <Icon icon={icon} className="TooltipEntity__icon" />
        {children}
        {showTooltipAt && (
          <Portal
            node={showTooltipAt.container}
            onClose={this.closeTooltip}
            closeOnClick
            closeOnType
            closeOnResize
          >
            <Tooltip target={showTooltipAt} direction="top">
              {(
                <a
                  href={url}
                  title={url}
                  target="_blank"
                  rel="noopener noreferrer"
                  className="Tooltip__link"
                >
                  {shortenLabel(label)}
                </a>
              )}

              <button
                className="button Tooltip__button"
                onClick={this.onEdit}
              >
                Edit
              </button>

              <button
                className="button button-secondary no Tooltip__button"
                onClick={this.onRemove}
              >
                Remove
              </button>
            </Tooltip>
          </Portal>
        )}
      </a>
    );
  }
}

TooltipEntity.propTypes = {
  entityKey: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  onEdit: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired,
  icon: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.object.isRequired,
  ]).isRequired,
  label: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};

export default TooltipEntity;
