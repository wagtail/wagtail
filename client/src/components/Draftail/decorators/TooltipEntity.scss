$icon-size: 1em;

.TooltipEntity {
  cursor: pointer;

  &:hover {
    color: theme('colors.surface-button-default');
  }

  .icon-warning {
    color: theme('colors.critical.200');
  }

  .icon {
    width: $icon-size;
    height: $icon-size;
    margin-inline-end: 0.2em;
    vertical-align: middle;
  }
}
