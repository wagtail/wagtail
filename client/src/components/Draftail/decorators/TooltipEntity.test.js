import React from 'react';
import { shallow } from 'enzyme';

import TooltipEntity from './TooltipEntity';

describe('TooltipEntity', () => {
  it('works', () => {
    expect(shallow((
      <TooltipEntity
        entityKey="1"
        onEdit={() => {}}
        onRemove={() => {}}
        icon="#icon-test"
        url="https://www.example.com/"
        label="www.example.com"
      >
        test
      </TooltipEntity>
    ))).toMatchSnapshot();
  });

  it('shortened label', () => {
    expect(shallow((
      <TooltipEntity
        entityKey="1"
        onEdit={() => {}}
        onRemove={() => {}}
        icon="#icon-test"
        url="https://www.example.com/"
        label="www.example.example.example.com"
      >
        test
      </TooltipEntity>
    )).setState({
      showTooltipAt: document.createElement('div').getBoundingClientRect(),
    }).find('Tooltip a')
      .text()).toBe('www.example.example.â€¦');
  });

  it('#openTooltip', () => {
    const wrapper = shallow((
      <TooltipEntity
        entityKey="1"
        onEdit={() => {}}
        onRemove={() => {}}
        icon="#icon-test"
        url="https://www.example.com/"
        label="www.example.com"
      >
        test
      </TooltipEntity>
    ));

    const target = document.createElement('div');
    target.setAttribute('data-draftail-trigger', true);
    document.body.appendChild(target);
    document.body.setAttribute('data-draftail-editor-wrapper', true);

    wrapper.find('.TooltipEntity').simulate('mouseup', {
      target: target,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('#closeTooltip', () => {
    const wrapper = shallow((
      <TooltipEntity
        entityKey="1"
        onEdit={() => {}}
        onRemove={() => {}}
        icon="#icon-test"
        url="https://www.example.com/"
        label="www.example.com"
      >
        test
      </TooltipEntity>
    ));

    wrapper.find('.TooltipEntity').simulate('mouseup', {
      target: document.createElement('div'),
    });

    wrapper.instance().closeTooltip();

    expect(wrapper.state()).toEqual({
      showTooltipAt: null,
    });
  });

  it('#onEdit', () => {
    const onEdit = jest.fn();

    const wrapper = shallow((
      <TooltipEntity
        entityKey="1"
        onEdit={onEdit}
        onRemove={() => {}}
        icon="#icon-test"
        url="https://www.example.com/"
        label="www.example.com"
      >
        test
      </TooltipEntity>
    ));

    wrapper.instance().onEdit(new Event('click'));

    expect(onEdit).toHaveBeenCalled();
  });

  it('#onRemove', () => {
    const onRemove = jest.fn();

    const wrapper = shallow((
      <TooltipEntity
        entityKey="1"
        onEdit={() => {}}
        onRemove={onRemove}
        icon="#icon-test"
        url="https://www.example.com/"
        label="www.example.com"
      >
        test
      </TooltipEntity>
    ));

    wrapper.instance().onRemove(new Event('click'));

    expect(onRemove).toHaveBeenCalled();
  });
});
