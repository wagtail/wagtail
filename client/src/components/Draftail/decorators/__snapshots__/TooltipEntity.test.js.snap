// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipEntity #openTooltip 1`] = `
<a
  className="TooltipEntity"
  data-draftail-trigger={true}
  onMouseUp={[Function]}
  role="button"
>
  <Icon
    className="TooltipEntity__icon"
    icon="#icon-test"
    title={null}
  />
  test
  <Portal
    closeOnClick={true}
    closeOnResize={true}
    closeOnType={true}
    node={
      <body
        data-draftail-editor-wrapper="true"
      >
        <div
          data-draftail-trigger="true"
        />
      </body>
    }
    onClose={[Function]}
  >
    <Tooltip
      direction="top"
      target={
        Object {
          "container": <body
            data-draftail-editor-wrapper="true"
          >
            <div
              data-draftail-trigger="true"
            />
          </body>,
          "height": 0,
          "left": 0,
          "top": 0,
          "width": 0,
        }
      }
    >
      <a
        className="Tooltip__link"
        href="https://www.example.com/"
        rel="noopener noreferrer"
        target="_blank"
        title="https://www.example.com/"
      >
        www.example.com
      </a>
      <button
        className="button Tooltip__button"
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        className="button button-secondary no Tooltip__button"
        onClick={[Function]}
      >
        Remove
      </button>
    </Tooltip>
  </Portal>
</a>
`;

exports[`TooltipEntity works 1`] = `
<a
  className="TooltipEntity"
  data-draftail-trigger={true}
  onMouseUp={[Function]}
  role="button"
>
  <Icon
    className="TooltipEntity__icon"
    icon="#icon-test"
    title={null}
  />
  test
</a>
`;
